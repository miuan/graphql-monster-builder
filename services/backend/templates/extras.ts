import * as bcrypt from 'bcrypt-nodejs';
import * as jwt from 'jsonwebtoken';
import * as _ from 'lodash'

export const generateHash = function (password) {
  return bcrypt.hashSync(password, bcrypt.genSaltSync(8), null);
};

/**
 *
 * @param user
 * @param options
 */
export const generateTokenJWT = (tokenizeData, opts?): string => {
  let options = opts || {};

  if (!options) {
    options = {
      /* expires in 365 days */
      expiresIn: '365d',
    };
  }

  const tokenJWT = jwt.sign(tokenizeData, process.env.JWT_TOKEN_SECRET || 'protectql_test_secret', options);

  return tokenJWT;
};

export const genPasswordAndTokens = (userData) => {
  // TODO: schort the expire date
  const options = { expiresIn: '365d' };
  userData.token = generateTokenJWT({ id: userData.id, role: userData.role }, options);
  userData.refreshToken = generateTokenJWT({ id: userData.id }, options);
};


export const checkPasswordIsNotIncluded = (userData) => {
  if (userData.password) {
    throw 'User can change password only in userChangePasswordMutation(old, new)';
  }
};

export const generateLogin = (entry) => {
  return async (root, data, ctx) => {
    
    const user = await entry.models['user'].findOne({ email: data.email });

    if (user && bcrypt.compareSync(data.password, user._password)) {
      return user;
    }

    throw 'Email or password is not correct';
  };
};

export const generateParentLogin = (entry) => async (ctx) => {
  if(ctx.params.parentAccessToken != process.env.PARENT_ACCESS_TOKEN
    || ctx.params.parentUserId != process.env.PARENT_ACCESS_USER_ID){
    throw 'Unknown parent access token'
  }
  
  const adminEmail = process.env.ADMIN_EMAIL

  const user = await entry.models['user'].findOne({ email: adminEmail })
  if(!user.token) {
    genPasswordAndTokens(user)
    await user.save()
  }
  ctx.body = {token: user.token}
}

export const generateChangePassword = (entry) => {
  return async (root, data, ctx) => {
    const userModel = entry.models['user'];
    const user = await userModel.findById(data.userId);

    console.log(data, user);
    if (!user) {
      throw `Unknown user with id: ${data.userId}`;
    }

    if (!bcrypt.compareSync(data.oldPassword, user._password)) {
      throw `Old password is different than is currently used`;
    }

    const updatedUser = await userModel.findByIdAndUpdate(
      data.userId, 
      { _password:generateHash(data.newPassword) }, 
      { new: true },
    );

    return updatedUser;
  };
};

export const generateProtections = (entry, modelName) => {
  return {
    public: () => true,
    user: (ctx) => {
      if (ctx && ctx.state && ctx.state.user) {
        return true;
      }
    },
    owner: async (ctx, data, ownerField = 'user') => {
      if (ctx && ctx.state && ctx.state.user) {
        // NOTE: in edit, if `fd` contain also other fields, what they are edited
        //       then `await entry.models[modelName].findOne(fd)` not find anything
        //       because data are not corespond with what is in DB
        // const fd = {
        //   ...data
        // }
        // if(fd.id) {
        //   fd._id = fd.id
        //   delete fd.id
        // }
        // const u = await entry.models[modelName].findOne(fd, ownerField)
        
        const u = await entry.models[modelName].findById(data.id || data._id, ownerField)
        console.log('OWNER CHECK : user1 >>', {u, fd:(data.id || data._id), state: ctx.state, modelName, models: entry.models})
        console.log('u && u[ownerField] == ctx.state.user.id', u && u[ownerField] == ctx.state.user.id)
        return u && u[ownerField] == ctx.state.user.id;
      } else {
        return false
      }
    },
    role: async (ctx, roles: string[]) => {
      if (ctx && ctx.state && ctx.state.user) {
        const ctxUser = ctx.state.user;
        const userRoleModel = entry.models['userRole'];
      
        const userRole = await userRoleModel.findOne({ role: roles[0], users: ctxUser.id });
      
        return userRole != null;
      }
    },
    filter: async (ctx, data, filters: any[], roles: string[]) => {
      if (ctx && ctx.state && ctx.state.user) {
        const ctxUser = ctx.state.user;
        const dataFilter = data.filter || {}
 
        // if is something restricted to filter can't be filter behind OR
        // because it will pass security 
        // but show even data what is under protection
        if(dataFilter.OR) {
          return false
        }

        // TODO: INPORTANT! is catch if secure filter is behind one $and example: filter: {$and:[{user: {{userId}}}]}
        //       but not catch if is under depper $and ... example filter: {$and:[{$and:[{user: {{userId}}}]]}
        let AND = null
        if(dataFilter.AND) {
          AND = dataFilter.AND
        } else if(dataFilter.length) {
          AND = dataFilter
        } else {
          AND = [dataFilter]
        }

        for(let a of AND) {
          for(let filter of filters) {
            if(filter.name) {
              const value = _.get(a, filter.name)
              console.log('filter', filter.name, value)

              if(filter.value == value || (filter.value == '{{userId}}' && value == ctxUser.id)) {
                return true
              }
              // console.log('data.filter.AND', data.filter.AND)
              // console.log('data.filter.AND[0]', data.filter.AND[0])
              // console.log('data.filter.AND[0].user_every', data.filter.AND[0].user_every)
            }
          }
        }
        
      }
      return false
    },
    checkDataContainProtectedFields
  };
};

// ******************************
// * * *    F I L T E R     * * *
// ******************************

const filterANDOR = (filter, operator) => {
  if(filter.length == 1){
    return filterGen(filter[0])
  }
  else if(filter.length > 1) {
    const obj = {}
    obj[operator] = []
    for(const f of filter) {
      obj[operator].push(filterGen(f))
    }
    return obj
  } else {
    return undefined
  }

}


export const filterGen = (filter) => {

  console.log('DEBUG:FILTER', filter)
  let obj = {}

  if(!filter){
    return obj
  } else if(filter.AND) {
    return filterANDOR(filter.AND, '$and')
  } else if(filter.OR) {
    return filterANDOR(filter.OR, '$or')
  } 

  if(Object.keys(filter).length > 0) {
    for(const f in filter){
      const ff = filter[f]

      if(/_every$/.test(f)){
        obj[f.substr(0, f.length-6)] = filterGen(ff)
      } else if(/_none$/.test(f)){
        obj[f.substr(0, f.length-5)] = {'$not': filterGen(ff)}
      } else if(/_not_contains$/.test(f)){
        obj[f.substr(0, f.length-13)] = {$not: { "$regex": ff, "$options": "i" }}
      } else if(/_contains$/.test(f)){
        obj[f.substr(0, f.length-9)] = { "$regex": ff, "$options": "i" }
      } else if(/_not$/.test(f)){
        obj[f.substr(0, f.length-4)] = { $ne: ff }
      } else if(/_not_in$/.test(f)){
        obj[f.substr(0, f.length-7)] = { $not: { $in: ff } }
      } else if(/_in$/.test(f)){
        obj[f.substr(0, f.length-3)] = { $in: ff }
      } else if(/_not_starts_with$/.test(f)){
        obj[f.substr(0, f.length-16)] = {$not: { "$regex": `^${ff}`, "$options": "i" }}
      } else if(/_starts_with$/.test(f)){
        obj[f.substr(0, f.length-12)] = { "$regex": `^${ff}`, "$options": "i" }
      } else if(/_not_ends_with$/.test(f)){
        obj[f.substr(0, f.length-14)] = {$not: { "$regex": `${ff}$`, "$options": "i" }}
      } else if(/_ends_with$/.test(f)){
        obj[f.substr(0, f.length-10)] = { "$regex": `${ff}$`, "$options": "i" }
      } else if(/_lt$/.test(f)){
        obj[f.substr(0, f.length-3)] = { "$lt": ff }
      } else if(/_lt$/.test(f)){
        obj[f.substr(0, f.length-3)] = { "$lt": ff }
      } else if(/_lte$/.test(f)){
        obj[f.substr(0, f.length-4)] = { "$lte": ff }
      } else if(/_gt$/.test(f)){
        obj[f.substr(0, f.length-3)] = { "$gt": ff }
      } else if(/_gte$/.test(f)){
        obj[f.substr(0, f.length-4)] = { "$gte": ff }
      } else if(f == 'id') {
        obj['_id'] = ff
      } else if(typeof ff == 'string' || typeof ff == 'number') {
        obj = {...filter}
      } else {
        // TODO: stop going deeper than just object.id
        //       mongoose can't query populated fields
        //       https://stackoverflow.com/questions/17535587/mongoose-query-a-populated-field
        //       posible show message is funciton of prepaid version
        return filterGen(ff)
      }
    }
  } else {
    return filter
  }
  
  return obj
}

class ReachProtectedFields extends Error {
  fields: any

  constructor(protectedFields) {
    super(`Reach protected fields`)
    this.fields = protectedFields
  }
}

/**
 * should return any field with name starts double underscore example  `__port`, `__readolny`
 * 
 * @param data - data to check
 * @param path - actual path in data
 */
export const checkDataContainProtectedFields = (data, path='/') => {

  let founded = []
  const keys = Object.keys(data)
  if(keys.length > 0){
    for(const fieldName of keys){
      if(/^__/.test(fieldName) || /^_/.test(fieldName)){
        founded.push({name: fieldName, path: path})
      } else if(typeof(data[fieldName]) != 'string') {
        let fundedInDeep = checkDataContainProtectedFields(data[fieldName], `${path}${fieldName}/`)
        founded.push(...fundedInDeep)
      }
    }
  }

  return founded
}

export const createUser = async (entry, email: string, password: string, rolesIds: string[] = []) => {
  const userModel = entry.models['user']
  const userService = entry.services['user']

  let user = await userModel.findOne({email: email})

  if(!user) {
    // call service instead of simple model
    // because we want also create relation between user <- and -> roles
    user = await userService.create({email: email, _password: password, rolesIds})
    console.log(`User with email: ${email} and with pass: ${password} [CREATED]`, user)
  } else {
    user = await userService.update({_password: password, rolesIds}, user.id)
    console.log(`Update pass (${password}) to already existed user with email: ${email}`, user)
  }

  return user
}

export const createRole = async (entry, role: string) => {
  const userRoleModel = entry.models['userRole']
  let userRole = await userRoleModel.findOne({role})
  if(!userRole){
    userRole = await userRoleModel.create({role})
    console.log(`Role with name: ${role} [CREATED]`, userRole)
  } else {
    console.log(`Role with name: ${role} already exist`, userRole)
  }
  
  return userRole
}