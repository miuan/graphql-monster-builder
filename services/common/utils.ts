import { SchemaModel, SchemaModelRelationType, SchemaModelMember } from './types'
import logger from '../log'
const log = logger.getLogger('utils')

export const searchModelsRelationsInModels = (relationName: string, models: SchemaModel[], exclude: string[] = []) => {
    for (const model of models) {
        if (exclude.indexOf(model.modelName) === -1) {
            for (const member of model.members) {
                if (member.relation && member.relation.name === relationName) {
                    return { model, member }
                }
            }
        }
    }

    return null
}

// in many to many relation we need create only one add or delete methods
// for example User and UserRole we want only userAddToUserRole and userRemoveFromUserRole
// but because the relation is cross each others it will create also userRoleAddToUser and userRoleRemoveFromUser
export const getOnlyOneRelatedMember = (member: SchemaModelMember) => {
    const {
        relation: { name: relationName, relatedMember },
    } = member

    if (member.isArray && (!relatedMember.isArray || member.modelName > relatedMember?.modelName)) {
        return member
    }

    // if any of them is not array, not take any
    return null
}

export const firstToLower = (name) => name.charAt(0).toLowerCase() + name.slice(1)
export const firstToUpper = (name) => name.charAt(0).toUpperCase() + name.slice(1)
