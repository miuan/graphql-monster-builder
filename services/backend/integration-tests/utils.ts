import * as fs from 'fs'
import * as service from '../generators/service'
import * as _ from 'lodash'
import * as path from 'path'
import * as bcrypt from 'bcrypt'
import { SchemaModelRelationType } from '../../common/types'
import { exportAsFromString } from '../build'
import { createTestClient } from 'apollo-server-testing'
import gql from 'graphql-tag'
import * as mongoose from 'mongoose'

import * as request from 'supertest'

export const TEST_DIR = 'templates/__generated_services_for_test__'
const SERVICE_BACKEND = './services/backend/'
export const testDirFullPath = `${SERVICE_BACKEND}${TEST_DIR}`

export function prepareDirectories() {
    if (!fs.existsSync(testDirFullPath)) {
        fs.mkdirSync(testDirFullPath)
    }

    if (!fs.existsSync(`${testDirFullPath}/services`)) {
        fs.mkdirSync(`${testDirFullPath}/services`)
    }

    if (!fs.existsSync(`${testDirFullPath}/models`)) {
        fs.mkdirSync(`${testDirFullPath}/models`)
    }

    //if(!fs.existsSync(`${testDirFullPath}/extras.ts`)){
    fs.writeFileSync(
        `${testDirFullPath}/extras.ts`,
        `
export const filterGen = jest.fn()
export default filterGen
`,
    )
}

export const createService = (model) => {
    const created = service.createService(model)
    const camelName = _.camelCase(model.modelName)
    fs.writeFileSync(`${testDirFullPath}/services/${camelName}Service.ts`, created)
    fs.writeFileSync(
        `${testDirFullPath}/models/${camelName}.ts`,
        `export const ${camelName}Model = {
        create: jest.fn(), //.mockImplementation(async (data)=>data),
        findByIdAndUpdate: jest.fn(), //.mockImplementation(async (id, data)=>data),
        findByIdAndRemove: jest.fn() //.mockImplementation(async (data)=>data),
    } as any;
    export const Types = {ObjectId: jest.fn().mockImplementation(()=>'random-id')}
    export default ${camelName}Model`,
    )

    try {
        const service = require(`../${TEST_DIR}/services/${camelName}Service.ts`)
        const model = require(`../${TEST_DIR}/models/${camelName}.ts`)[`${camelName}Model`]
        return {
            service,
            model,
            mockClearToAll: () => {
                model.create.mockClear()
                model.findByIdAndUpdate.mockClear()
                model.findByIdAndRemove.mockClear()
            },
        }
    } catch (ex) {
        console.error(ex)
        throw ex
    }
}

export function createServiceWithRelation(meType, required) {
    let relationService
    const modelName = _.camelCase(`Relation ${meType} ${required} `)
    const memberWeWithRelationOthers = {
        name: 'toThem',
        relation: {
            name: 'WeConnectToOthers',
            isRequired: false,
            type: meType,
            relatedModel: {
                modelName: 'ModelWithOthers',
                members: [
                    {
                        name: 'backToMe',
                        isRequired: required === 'required',
                        relation: {
                            name: 'WeConnectToOthers',
                            type: meType,
                        },
                    },
                ],
            },
        },
    }

    const updateMany = jest.fn()
    const findOne = jest.fn()
    const remove = jest.fn()
    const create = jest.fn()

    const entry = {
        hooks: { services: {} },
        services: {
            modelWithOthers: {
                create,
                remove,
            },
        },
        models: {
            modelWithOthers: {
                findOne,
                updateMany,
            },
        },
    }
    relationService = createService({ modelName, members: [memberWeWithRelationOthers] })
    // relationService.model.findByIdAndUpdate.mockResolvedValue({id: 'resolved-findByIdAndUpdate-id-1'})

    const creatGen = relationService.service[`${modelName}Create`]
    // expect(creatGen).toBeFunction()

    const createFn = creatGen(entry)
    //  expect(createFn).toBeFunction()

    const updateGen = relationService.service[`${modelName}Update`]
    // expect(updateGen).toBeFunction()

    const updateFn = updateGen(entry)
    // expect(updateFn).toBeFunction()

    const removeGen = relationService.service[`${modelName}Remove`]
    // expect(removeGen).toBeFunction()

    const removeFn = removeGen(entry)
    // expect(removeFn).toBeFunction()

    return {
        service: relationService,
        entry,
        updateFn,
        createFn,
        removeFn,
        mockClearToAll: () => {
            relationService.mockClearToAll()

            create.mockClear()
            findOne.mockClear()
            updateMany.mockClear()

            remove.mockClear()
        },
    }
}

// export function reverseType(meType){
//     if(meType == SchemaModelRelationType.MANY_TO_MANY) return SchemaModelRelationType.MANY_TO_MANY
//     if(meType == SchemaModelRelationType.MANY_TO_ONE) return SchemaModelRelationType.ONE_TO_MANY
//     if(meType == SchemaModelRelationType.ONE_TO_MANY) return SchemaModelRelationType.MANY_TO_ONE
//     else return SchemaModelRelationType.ONE_TO_ONE
// }

// NODE: database name 'test_integration_all' - is for protend exception with multiple call connect with different connect string
export async function generateAndRunServerFromSchema(name: string, schema: string, port = 3001, dbname = 'test_integration_all') {
    let server
    let dropCollectionPromises = []

    const integrationTestServerPath = path.resolve(`../__generated_servers_for_integration_test__/${name}`)

    await remakeServer(integrationTestServerPath, name, schema, port, dbname)

    try {
        const module = require(path.join(integrationTestServerPath, 'server'))
        server = await module.connectionPromise

        // NOTE:
        //    createTestClient - skip the requeste layer and skip the header part
        //    https://github.com/apollographql/apollo-server/issues/2277
        // const { query, mutate } = createTestClient(server.apollo)

        const query = async (body: { query: string; variables: any }, token?: string) => {
            const req = request(server.koa).post('/graphql').set('Content-Type', 'application/json').set('Accept', 'application/json')

            if (token) {
                req.set('Authorization', `Bearer ${token}`)
            }

            return (await req.send(body)).body
        }

        const mutate = async ({ mutation, variables }: { mutation: string; variables: any }, token?: string) => {
            return query({ query: mutation, variables }, token)
        }

        for (const modelName of Object.keys(await server.entry.models)) {
            await server.entry.models[modelName].remove({})
        }

        // await new Promise((resolve) => setTimeout(resolve, 1000))
        // upload user again
        await module.updateAdminUser(true)

        const count = await server.entry.models.user.count({})

        return { ...server, query, mutate }
    } catch (ex) {
        console.error(ex)
    }
}

async function remakeServer(integrationTestServerPath: string, name: string, schema: string, port: number, dbname) {
    fs.rmdirSync(integrationTestServerPath, { recursive: true })
    await exportAsFromString(name, schema, integrationTestServerPath, {
        server: {
            config: {
                "./gen/graphql.schema'": `../__generated_servers_for_integration_test__/${name}/gen/graphql.schema'`,
                "path: './config/environment'": `path: '../__generated_servers_for_integration_test__/${name}/config/environment'`,
            },
        },
        extras: {
            config: { "from './sendMail'": "from '../services/sendMail'" },
        },
    })

    const modules = path.resolve(`./node_modules`)

    fs.symlinkSync(modules, path.join(integrationTestServerPath, 'node_modules'))

    const envPath = path.join(integrationTestServerPath, 'config/environment/')
    fs.mkdirSync(envPath, { recursive: true })
    fs.writeFileSync(
        path.join(envPath, '.env'),
        `
ADMIN_EMAIL=admin@admin.test
ADMIN_PASSWORD=${bcrypt.hashSync('admin@admin.test', 1)}
    `,
    )

    const serverPaht = path.join(integrationTestServerPath, 'server.ts')
    await replaceInFile(serverPaht, (data) => {
        return data.replace(/3001/g, port).replace('db_graphql_monster', dbname)
    })
}

async function replaceInFile(file, replaceFn) {
    return new Promise((resolve) => {
        fs.readFile(file, 'utf8', function (err, data) {
            if (err) {
                return console.log(err)
            }

            const result = replaceFn(data) //data.replace(/string to be replaced/g, 'replacement')

            fs.writeFile(file, result, 'utf8', function (err) {
                if (err) return console.log(err)
                resolve(null)
            })
        })
    })
}

export async function disconnectFromServer(server: any) {
    await server.koa.close()
    await new Promise((resolve, reject) => {
        server.mongoDB.close(() => {
            resolve(1)
        })
    })
    console.log('server closed!')
}

export function loadGraphQL(fileName) {
    let file = fs.readFileSync(fileName).toString()
    if (file.startsWith('#import "')) {
        let fragment = '',
            fragment2 = ''
        const match = file.match(/#import "(.*?)"/)
        if (match?.length > 0) {
            const fragmenName = path.join(path.dirname(fileName), match[1])
            fragment = fs.readFileSync(fragmenName).toString()

            const match2 = fragment.match(/#import "(.*?)"/)
            if (match2?.length > 0) {
                const fragmenName2 = path.join(path.dirname(fileName), match2[1])
                fragment2 = fs.readFileSync(fragmenName2).toString()
            }
        }

        file = `
            ${fragment2}
            ${fragment}

            ${file}
        `
    }

    const ql = file //gql(file);
    return ql
}

export function jsonWithoutCircularStructure(circularJSON) {
    const replacerFunc = () => {
        const listOfObject = []
        return (key, value) => {
            if (typeof value === 'object' && value !== null) {
                if (listOfObject.includes(value)) {
                    return
                }
                listOfObject.push(value)
            }
            return value
        }
    }

    return JSON.parse(JSON.stringify(circularJSON, replacerFunc()))
}
