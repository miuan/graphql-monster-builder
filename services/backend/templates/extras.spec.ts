import * as extras from './extras'
import { mocked } from 'ts-jest/utils'

describe('extras', () => {
    it.todo('filterGen')
    it.todo('generateProtections: user')
    it.todo('generateProtections: owner')
    it.todo('generateProtections: role')
    it.todo('generateChangePassword')
    it.todo('generateLogin')
    it.todo('checkPasswordIsNotIncluded')
    it.todo('genPasswordAndTokens')
    it.todo('generateTokenJWT')
    it.todo('generateHash')
    describe('propertiesToArray', () => {
        it('1', () => {
            const pa = extras.propertiesToArray({
                AND: [
                    {
                        OR: [
                            {
                                some: { id: '1' },
                            },
                        ],
                    },
                    { AND: [{ user: { id: 'every' } }, { deer: { id: '1' } }] },
                ],
            })

            expect(pa).toEqual(['AND.0.OR.0.some.id', 'AND.1.AND.0.user.id', 'AND.1.AND.1.deer.id'])
        })
    })

    describe('filterValue', () => {
        it('find userId in root', () => {
            const filter = {
                user: { id: 'userId1' },
            }

            const { value, path, pathIs } = extras.filterValue(filter, 'user.id')
            expect(value).toEqual('userId1')
            expect(path).toEqual([])
            expect(pathIs).toEqual(extras.PATHIS.ALWAYS)
        })

        it('find userId after AND', () => {
            const filter = {
                AND: [
                    {
                        user: { id: 'userId1' },
                    },
                ],
            }

            const { value, path, pathIs } = extras.filterValue(filter, 'user.id')
            expect(value).toEqual('userId1')
            expect(path).toEqual(['AND'])
            expect(pathIs).toEqual(extras.PATHIS.ALWAYS)
        })

        it('find userId after AND AND', () => {
            const filter = {
                AND: [
                    {
                        AND: [
                            {
                                user: { id: 'userId1' },
                            },
                        ],
                    },
                ],
            }

            const { value, path, pathIs } = extras.filterValue(filter, 'user.id')
            expect(value).toEqual('userId1')
            expect(path).toEqual(['AND', 'AND'])
            expect(pathIs).toEqual(extras.PATHIS.ALWAYS)
        })

        it('find userId after OR AND', () => {
            const filter = {
                OR: [
                    {
                        AND: [
                            {
                                user: { id: 'userId1' },
                            },
                        ],
                    },
                ],
            }

            const { value, path, pathIs } = extras.filterValue(filter, 'user.id')
            expect(value).toEqual('userId1')
            expect(path).toEqual(['OR', 'AND'])
            expect(pathIs).toEqual(extras.PATHIS.SOME)
        })

        it('userId not present', () => {
            const filter = {
                OR: [
                    {
                        AND: [
                            {
                                noUserId: 'userId1',
                            },
                        ],
                    },
                ],
            }

            const { value, path, pathIs } = extras.filterValue(filter, 'user.id')
            expect(value).toEqual(null)
            expect(path).toEqual(null)
            expect(pathIs).toEqual(extras.PATHIS.NEVER)
        })

        it('find simple value', () => {
            const filter = {
                OR: [
                    {
                        AND: [
                            {
                                value: 'value-1',
                            },
                        ],
                    },
                ],
            }

            const { value, path, pathIs } = extras.filterValue(filter, 'value')
            expect(value).toEqual('value-1')
            expect(path).toEqual(['OR', 'AND'])
            expect(pathIs).toEqual(extras.PATHIS.SOME)
        })

        it('should not determine find value because is in filter twice', () => {
            const filter = {
                OR: [
                    {
                        AND: [
                            {
                                value: 'value-1',
                            },
                        ],
                    },
                    {
                        value: 'value-1',
                    },
                ],
            }

            const { value, path, pathIs } = extras.filterValue(filter, 'value')
            expect(value).toEqual(undefined)
            expect(path).toEqual(undefined)
            expect(pathIs).toEqual(extras.PATHIS.UNDETERMINED)
        })
    })
    describe('generateProtections: filter', () => {
        it('should return false because no user in ctx', async () => {
            const protection = extras.generateProtections({ models: { userRole: {} } }, 'model-name-test')
            expect(await protection.filter({}, {}, [{ userId: '{{userId}}' }], null)).toBeFalsy()
        })

        it('should return false because no filter', async () => {
            const protection = extras.generateProtections({ models: { userRole: {} } }, 'model-name-test')
            expect(await protection.filter({ state: { user: { id: 'user-id-1' } } }, {}, [{ userId: '{{userId}}' }], null)).toBeFalsy()
        })

        it('should return false because filter starts with OR', async () => {
            const protection = extras.generateProtections({ models: { userRole: {} } }, 'model-name-test')
            expect(await protection.filter({ state: { user: { id: 'user-id-1' } } }, { filter: [{ OR: [{ userId: '1' }] }] }, [{ userId: '{{userId}}' }], null)).toBeFalsy()
        })

        it('should return false because filter userId:1 is not same as allowed filter userId:{{userId}}', async () => {
            const protection = extras.generateProtections({ models: { userRole: {} } }, 'model-name-test')
            expect(await protection.filter({ state: { user: { id: 'user-id-1' } } }, { filter: [{ AND: [{ userId: '1' }] }] }, [{ userId: '{{userId}}' }], null)).toBeFalsy()
        })

        it('should return false because filter user.id is not same as context user (allowed filter userId:{{userId}}', async () => {
            const protection = extras.generateProtections({ models: { userRole: {} } }, 'model-name-test')
            expect(
                await protection.filter({ state: { user: { id: 'user-id-1' } } }, { filter: [{ user_every: { id: 'wrong-user-id' } }] }, [{ name: 'user_every.id', value: '{{userId}}' }], null),
            ).toBeFalsy()
        })

        it('should return TRUE because filter user.id is same as context user (allowed filter userId:{{userId}}', async () => {
            const protection = extras.generateProtections({ models: { userRole: {} } }, 'model-name-test')
            expect(
                await protection.filter({ state: { user: { id: 'user-id-1' } } }, { filter: [{ user_every: { id: 'user-id-1' } }] }, [{ name: 'user_every.id', value: '{{userId}}' }], null),
            ).toBeTruthy()
        })

        it('should return false because filter user.id is different than custom user id (wrong-user-id vs custom-user-id', async () => {
            const protection = extras.generateProtections({ models: { userRole: {} } }, 'model-name-test')
            expect(
                await protection.filter({ state: { user: { id: 'user-id-1' } } }, { filter: [{ user_every: { id: 'wrong-user-id' } }] }, [{ name: 'user_every.id', value: 'custom-user-id' }], null),
            ).toBeFalsy()
        })

        it('should return TRUE because filter user.id is same custom user id (allowed custom-user-id', async () => {
            const protection = extras.generateProtections({ models: { userRole: {} } }, 'model-name-test')
            expect(
                await protection.filter({ state: { user: { id: 'user-id-1' } } }, { filter: [{ user_every: { id: 'custom-user-id' } }] }, [{ name: 'user_every.id', value: 'custom-user-id' }], null),
            ).toBeTruthy()
        })
    })

    describe('generateParentLogin', () => {
        let genPasswordAndTokensStub = jest.spyOn(extras, 'genPasswordAndTokens')

        afterEach(() => {
            mocked(genPasswordAndTokensStub).mockClear()
        })

        afterAll(() => {
            mocked(extras.genPasswordAndTokens).mockRestore()
        })

        it('should pass', async () => {
            const saveStub = jest.fn()
            const findOneStub = jest.fn().mockImplementation(() => ({
                token: '1',
                save: saveStub,
            }))
            const sendStub = jest.fn()
            const genPasswordAndTokensStub = jest.spyOn(extras, 'genPasswordAndTokens').mockImplementation(() => {})

            process.env.PARENT_ACCESS_TOKEN = '7674123'
            process.env.PARENT_ACCESS_USER_ID = '4321123'
            process.env.ADMIN_EMAIL = 'admin@email.com'
            const parentLogin = extras.generateParentLogin({
                models: {
                    user: {
                        findOne: findOneStub,
                    },
                },
            })

            const ctx = {
                params: {
                    parentAccessToken: '7674123',
                    parentUserId: '4321123',
                },
                body: null,
            }

            await parentLogin(ctx)

            expect(findOneStub).toBeCalledWith({ email: 'admin@email.com' })
            expect(genPasswordAndTokensStub).not.toBeCalled()
            expect(ctx.body).toEqual({ token: '1' })
        })

        it('should pass and generate new tokens', async () => {
            const saveStub = jest.fn()
            const findOneStub = jest.fn().mockResolvedValue({
                save: saveStub,
            })
            const sendStub = jest.fn()
            genPasswordAndTokensStub.mockImplementation((user) => {
                user.token = 'genPasswordAndTokensStub-token-1'
            })

            process.env.PARENT_ACCESS_TOKEN = '7674123'
            process.env.PARENT_ACCESS_USER_ID = '4321123'
            process.env.ADMIN_EMAIL = 'admin@email.com'

            const parentLogin = extras.generateParentLogin({
                models: {
                    user: {
                        findOne: findOneStub,
                    },
                },
            })

            const ctx = {
                params: {
                    parentAccessToken: '7674123',
                    parentUserId: '4321123',
                },
                body: null,
            }

            await parentLogin(ctx)

            expect(findOneStub).toBeCalledWith({ email: 'admin@email.com' })
            expect(genPasswordAndTokensStub).toBeCalled()
            expect(ctx.body).toEqual({ token: 'genPasswordAndTokensStub-token-1' })
        })

        it('should fail wrong parent access token', async () => {
            const saveStub = jest.fn()
            const findOneStub = jest.fn().mockImplementation(() => ({
                token: '1',
                save: saveStub,
            }))
            const sendStub = jest.fn()
            const genPasswordAndTokensStub = jest.spyOn(extras, 'genPasswordAndTokens').mockImplementation(() => {})

            process.env.PARENT_ACCESS_TOKEN = '123'
            process.env.ADMIN_EMAIL = 'admin@email.com'
            const parentLogin = extras.generateParentLogin({
                models: {
                    user: {
                        findOne: findOneStub,
                    },
                },
            })

            await expect(
                parentLogin({
                    params: { parentAccessToken: 'wrong123' },
                    send: sendStub,
                }),
            ).rejects.toEqual('Unknown parent access token')

            expect(findOneStub).not.toBeCalled()
            expect(genPasswordAndTokensStub).not.toBeCalled()
            expect(sendStub).not.toBeCalled()
        })

        it('should fail wrong access user id', async () => {
            const saveStub = jest.fn()
            const findOneStub = jest.fn().mockImplementation(() => ({
                token: '1',
                save: saveStub,
            }))
            const sendStub = jest.fn()
            const genPasswordAndTokensStub = jest.spyOn(extras, 'genPasswordAndTokens').mockImplementation(() => {})

            process.env.PARENT_ACCESS_TOKEN = '123'
            process.env.PARENT_ACCESS_USER_ID = '123'
            process.env.ADMIN_EMAIL = 'admin@email.com'
            const parentLogin = extras.generateParentLogin({
                models: {
                    user: {
                        findOne: findOneStub,
                    },
                },
            })

            await expect(
                parentLogin({
                    params: {
                        parentAccessToken: '123',
                        parentUserId: 'wrong123',
                    },
                    send: sendStub,
                }),
            ).rejects.toEqual('Unknown parent access token')

            expect(findOneStub).not.toBeCalled()
            expect(genPasswordAndTokensStub).not.toBeCalled()
            expect(sendStub).not.toBeCalled()
        })
    })

    describe('checkDataContainProtectedFields', () => {
        it('should pass no issues with checkDataContainProtectedFields', () => {
            const found = extras.checkDataContainProtectedFields({
                name: 'name1',
                personal: {
                    year: 2001,
                    contacts: {
                        tel: 'phone-number',
                    },
                },
            })

            expect(found).toHaveLength(0)
        })

        it('should pass no issues with checkDataContainProtectedFields', () => {
            const found = extras.checkDataContainProtectedFields({
                name: 'name1',
                __private: 3,
                __private_text: 'who ho ho',
                personal: {
                    year: 2001,
                    __unreachable: 'also not editable',
                    contacts: {
                        tel: 'phone-number',
                        __connect_code: 'also not editable',
                    },
                },
            })

            expect(found).toHaveLength(4)
            expect(found).toMatchSnapshot()
        })
    })
})
