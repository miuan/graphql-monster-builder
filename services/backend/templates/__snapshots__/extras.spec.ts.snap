// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkDataContainProtectedFields should pass no issues with checkDataContainProtectedFields 1`] = `
Array [
  Object {
    "name": "__private",
    "path": "/",
  },
  Object {
    "name": "__private_text",
    "path": "/",
  },
  Object {
    "name": "__unreachable",
    "path": "/personal/",
  },
  Object {
    "name": "__connect_code",
    "path": "/personal/contacts/",
  },
]
`;

exports[`extras checkDataContainProtectedFields should pass no issues with checkDataContainProtectedFields 1`] = `
Array [
  Object {
    "name": "__private",
    "path": "/",
  },
  Object {
    "name": "__private_text",
    "path": "/",
  },
  Object {
    "name": "__unreachable",
    "path": "/personal/",
  },
  Object {
    "name": "__connect_code",
    "path": "/personal/contacts/",
  },
]
`;
